File: swift/SDTrustKit/Sources/SDTrustKit/ValidationReport.swift
import Foundation

public struct ValidationReport: Decodable, Equatable, Sendable {
    public let steps: [ValidationStep]
    public let signerName: String?
    public let signerNames: [String]
    public let signingTime: String?
    public let verdict: ValidationVerdict
    public let signatures: [SignatureReport]
    public let documentTimestamps: [SignatureReport]
    public let standards: StandardsValidationResult
}

public struct SignatureReport: Decodable, Equatable, Sendable {
    public let index: Int
    public let total: Int
    public let signedRevisionSize: Int
    public let currentFileSize: Int
    public let byteRange: [Int]
    public let steps: [ValidationStep]
    public let signerName: String?
    public let signingTime: String?
    public let signerCertificate: CertificateDetails?
    public let certificateChain: [CertificateDetails]
    public let timestampDetails: TimestampDetails?
    public let verdict: ValidationVerdict
}

public struct ValidationStep: Decodable, Equatable, Sendable {
    public let kind: StepKind
    public let name: String
    public let status: StepStatus
    public let detail: String
}

public struct StandardsValidationResult: Decodable, Equatable, Sendable {
    public let indication: ValidationIndication
    public let subIndication: ValidationSubIndication
    public let diagnostic: String?
}

public struct CertificateDetails: Decodable, Equatable, Sendable {
    public let der: [UInt8]?
    public let subjectSummary: String
    public let issuerSummary: String
    public let commonName: String?
    public let serialNumberHex: String
    public let sha1Fingerprint: String
    public let sha256Fingerprint: String
}

public struct TimestampDetails: Decodable, Equatable, Sendable {
    public let timestampTime: String?
    public let policyOID: String?
    public let serialNumberHex: String?
    public let messageImprintAlgorithm: String?
    public let messageImprintHash: String?
    public let tsaCertificate: CertificateDetails?
    public let tsaCertificateChain: [CertificateDetails]
    public let trustDetail: String?
}

public enum StepStatus: String, Decodable, Sendable {
    case ok
    case fail
    case warn
    case skip
}

public enum ValidationVerdict: String, Decodable, Sendable {
    case error
    case valid
    case warning
    case inconclusive
    case invalid
    case noSignatures
}

public enum ValidationIndication: String, Decodable, Sendable {
    case passed
    case failed
    case needsEvidence
}

public enum ValidationSubIndication: String, Decodable, Sendable {
    case none
    case formatIssue
    case documentModifiedAfterSigning
    case documentHashMismatch
    case signatureCryptographyIssue
    case signingCertificateMissing
    case certificateChainIssue
    case revocationEvidenceUnavailable
    case certificateRevoked
    case timestampEvidenceIssue
    case cryptographicConstraintIssue
}

public enum StepKind: String, Decodable, Sendable {
    case parsePDF
    case signatureFieldResolution
    case byteRangeCoverage
    case byteRangeBounds
    case documentModifiedAfterSigning
    case cmsStructure
    case padesBaselineRequirements
    case signerInfoPresent
    case messageDigestMatches
    case messageDigestAttribute
    case signerCertificatePresent
    case signerCertificateValidity
    case signerCertificateKeyUsage
    case signerCertificateExtendedKeyUsage
    case signatureVerifySignedAttributes
    case signatureVerifyContent
    case signerCertificateChain
    case tsaTimestamp
    case tsaMessageImprint
    case tsaSignatureVerify
    case tsaExtendedKeyUsage
    case tsaCertificateChain
    case documentTimestamp
    case revocationSigner
    case other
}

extension JSONDecoder {
    static var rustReport: JSONDecoder {
        JSONDecoder()
    }
}

File: kotlin/SDTrustKit/sdtrustkit/src/main/java/com/sdtrustkit/ValidationReport.kt
package com.sdtrustkit

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class ValidationReport(
    val steps: List<ValidationStep>,
    val signerName: String? = null,
    val signerNames: List<String> = emptyList(),
    val signingTime: String? = null,
    val verdict: ValidationVerdict,
    val signatures: List<SignatureReport> = emptyList(),
    val documentTimestamps: List<SignatureReport> = emptyList(),
    val standards: StandardsValidationResult,
    val padesLevel: PadesLevel,
    val preservation: PreservationAssessment,
)

@Serializable
data class SignatureReport(
    val index: Int,
    val total: Int,
    val signedRevisionSize: Int,
    val currentFileSize: Int,
    val byteRange: List<Int>,
    val steps: List<ValidationStep>,
    val signerName: String? = null,
    val signingTime: String? = null,
    val signerCertificate: CertificateDetails? = null,
    val certificateChain: List<CertificateDetails> = emptyList(),
    val timestampDetails: TimestampDetails? = null,
    val verdict: ValidationVerdict,
    val padesLevel: PadesLevel,
    val preservation: PreservationAssessment,
)

@Serializable
data class ValidationStep(
    val kind: StepKind,
    val name: String,
    val status: StepStatus,
    val detail: String,
)

@Serializable
data class StandardsValidationResult(
    val indication: ValidationIndication,
    val subIndication: ValidationSubIndication,
    val diagnostic: String? = null,
)

@Serializable
data class CertificateDetails(
    val der: List<Int>? = null,
    val subjectSummary: String,
    val issuerSummary: String,
    val commonName: String? = null,
    val serialNumberHex: String,
    val sha1Fingerprint: String,
    val sha256Fingerprint: String,
)

@Serializable
data class TimestampDetails(
    val timestampTime: String? = null,
    val policyOID: String? = null,
    val serialNumberHex: String? = null,
    val messageImprintAlgorithm: String? = null,
    val messageImprintHash: String? = null,
    val tsaCertificate: CertificateDetails? = null,
    val tsaCertificateChain: List<CertificateDetails> = emptyList(),
    val trustDetail: String? = null,
)

@Serializable
data class PreservationAssessment(
    val level: PreservationLevel,
    val label: String,
    val detail: String,
)

@Serializable
enum class PadesLevel {
    @SerialName("unknown")
    Unknown,

    @SerialName("baselineB")
    BaselineB,

    @SerialName("baselineT")
    BaselineT,

    @SerialName("baselineLT")
    BaselineLT,

    @SerialName("baselineLTA")
    BaselineLTA,
}

@Serializable
enum class PreservationLevel {
    @SerialName("unknown")
    Unknown,

    @SerialName("basic")
    Basic,

    @SerialName("timestamped")
    Timestamped,

    @SerialName("longTerm")
    LongTerm,

    @SerialName("archival")
    Archival,
}

@Serializable
enum class StepStatus {
    @SerialName("ok")
    Ok,

    @SerialName("fail")
    Fail,

    @SerialName("warn")
    Warn,

    @SerialName("skip")
    Skip,
}

@Serializable
enum class ValidationVerdict {
    @SerialName("error")
    Error,

    @SerialName("valid")
    Valid,

    @SerialName("warning")
    Warning,

    @SerialName("inconclusive")
    Inconclusive,

    @SerialName("invalid")
    Invalid,

    @SerialName("noSignatures")
    NoSignatures,
}

@Serializable
enum class ValidationIndication {
    @SerialName("passed")
    TotalPassed,

    @SerialName("failed")
    TotalFailed,

    @SerialName("needsEvidence")
    Indeterminate,
}

@Serializable
enum class ValidationSubIndication {
    @SerialName("none")
    None,

    @SerialName("formatIssue")
    FormatFailure,

    @SerialName("documentModifiedAfterSigning")
    DocumentModifiedAfterSigning,

    @SerialName("documentHashMismatch")
    HashFailure,

    @SerialName("signatureCryptographyIssue")
    SignatureCryptoFailure,

    @SerialName("signingCertificateMissing")
    SigningCertificateNotFound,

    @SerialName("certificateChainIssue")
    CertificateChainGeneralFailure,

    @SerialName("revocationEvidenceUnavailable")
    RevocationOutOfBoundsNoPoe,

    @SerialName("certificateRevoked")
    Revoked,

    @SerialName("timestampEvidenceIssue")
    TimestampGeneralFailure,

    @SerialName("cryptographicConstraintIssue")
    CryptographicConstraintsFailure,
}

@Serializable
enum class StepKind {
    @SerialName("parsePDF")
    ParsePDF,

    @SerialName("signatureFieldResolution")
    SignatureFieldResolution,

    @SerialName("byteRangeCoverage")
    ByteRangeCoverage,

    @SerialName("byteRangeBounds")
    ByteRangeBounds,

    @SerialName("documentModifiedAfterSigning")
    DocumentModifiedAfterSigning,

    @SerialName("cmsStructure")
    CmsStructure,

    @SerialName("padesBaselineRequirements")
    PadesBaselineRequirements,

    @SerialName("signerInfoPresent")
    SignerInfoPresent,

    @SerialName("messageDigestMatches")
    MessageDigestMatches,

    @SerialName("messageDigestAttribute")
    MessageDigestAttribute,

    @SerialName("signerCertificatePresent")
    SignerCertificatePresent,

    @SerialName("signerCertificateValidity")
    SignerCertificateValidity,

    @SerialName("signerCertificateKeyUsage")
    SignerCertificateKeyUsage,

    @SerialName("signerCertificateExtendedKeyUsage")
    SignerCertificateExtendedKeyUsage,

    @SerialName("signatureVerifySignedAttributes")
    SignatureVerifySignedAttributes,

    @SerialName("signatureVerifyContent")
    SignatureVerifyContent,

    @SerialName("signerCertificateChain")
    SignerCertificateChain,

    @SerialName("tsaTimestamp")
    TsaTimestamp,

    @SerialName("tsaMessageImprint")
    TsaMessageImprint,

    @SerialName("tsaSignatureVerify")
    TsaSignatureVerify,

    @SerialName("tsaExtendedKeyUsage")
    TsaExtendedKeyUsage,

    @SerialName("tsaCertificateChain")
    TsaCertificateChain,

    @SerialName("documentTimestamp")
    DocumentTimestamp,

    @SerialName("revocationSigner")
    RevocationSigner,

    @SerialName("other")
    Other,
}
